### ISSUE TITLE
Alpha-Beta search algorithm

### ISSUE THREAD

**Issue by Dohbedoh** [ChessCorp/chess-ai-kong#1]
##### Minimax
The algorithm relies on the **Minimax** algorithm, a 2 agents algorithm in which one agent (the AI) try to maximize the evaluation and the other agent (the opponent) try to minimize it. Following is the basic algorithm:

```
function minimax(board, depth, maximizing)
    if depth = 0
        return board evaluation
    if maximizing
        score = -∞
        for each move
            updatedBoard= apply move
            val = minimax(updatedBoard, depth - 1, false)
            score = max(score, val)
        return score
    else
        score = +∞
        for each move
            val = minimax(updatedBoard, depth - 1, true)
            score= min(score, val)
        return score
```

##### Alpha Beta Pruning
The alpha-beta pruning algorithm introduce alpha and beta variables which represents respectively the AI and the opponent best options (instead of using -∞ and +∞ in Minimax). Following is the basic algorithm:

```
function alphaBetaMax(board, depth, α, β,)
    if depth = 0
        return board evaluation
    for each move
        updatedBoard = apply move
        score = alphaBetaMin( updatedBoard, α, β, depthleft - 1 );
        if( score >= β)
           //β cut-off
           return β
        if( score > α)
           //new max score
           α = score
    return α
 
function alphaBetaMin(board, depth, α, β)
    if depth = 0
        return board evaluation
    for each move
        updatedBoard = apply move
        score = alphaBetaMin( updatedBoard, α, β, depth- 1 );
        if( score <= α)
           //α cut-off
           return α
        if( score < β)
           //new min score
           β = score
    return β
```

The idea of the algorithm is that **when you find a good move, you can eliminate those who lead to a bad evaluation**. 

An extension can and should to be developed to enhance the algorithm: a function to sort the possible move (the good first to be able to eliminate the bad as fast as possible). As a matter of fact, many other techniques can be added to enhance the algorithm or to enhance the AI evaluation as well and will be created as new issues.

**Comment by Dohbedoh** [comment]
The algorithm is fully integrated. More tests should be written however before making any optimization.

**Comment by Dohbedoh** [comment]
Closed by 0.2.1.
